src/pid/environ.rs

```diff
@@ -0,0 +1,63 @@
+//! Process initial environment from `/proc/[pid]/environ`.
+
+use std::collections::BTreeMap;
+use std::fs::{File, metadata};
+use std::io::{Result, Read};
+use std::path::Path;
+
+use parsers::map_result;
+use nom::IResult;
+
+use libc::pid_t;
+
+/// Parses the provided buffer.
+fn parse_environ(input: &[u8]) -> IResult<&[u8], BTreeMap<Vec<u8>, Vec<u8>>> {
+    // Parse 'key=value" pair.
+    named!(
+        pair<&[u8], (&[u8], &[u8])>,
+        tuple!(take_until_and_consume!("="), take_until!("\0"))
+    );
+    // Parse (key=value\0) list.
+    named!(
+        multi<&[u8],Vec<(&[u8], &[u8])> >,
+        terminated!(
+            separated_list!(tag!("\0"), pair),
+            tag!("\0")
+        )
+    );
+    multi(input).map(|r| {
+        r.into_iter()
+            .map(|(x, y)| (x.to_vec(), y.to_vec()))
+            .collect()
+    })
+}
+
+#[test]
+fn test_parse() {
+    use nom::IResult::Done;
+    let res = parse_environ(&b"key1=val1\0key2=val2\0"[..]);
+    let reference = vec![
+        (b"key1".to_vec(), b"val1".to_vec()),
+        (b"key2".to_vec(), b"val2".to_vec()),
+    ];
+    assert_eq!(res, Done(&b""[..], reference.into_iter().collect()));
+}
+
+/// Parses the provided environ file.
+fn environ_path<P: AsRef<Path>>(path: P) -> Result<BTreeMap<Vec<u8>, Vec<u8>>> {
+    let file_size = metadata(&path)?.len();
+    let mut buf = vec![0u8; file_size as usize];
```

Comment:
This is racy with updates to the env, right?  I think it would be better and simpler just to use `File::read_to_end`.

Reply:
From `man 5 proc`:
```
If, after an execve(2), the process modifies its environment (e.g., by calling functions such as putenv(3) or modifying the environ(7) variable directly), this file  will not reflect those changes.
```

So the only race condition I can think of is the process is closed in between getting the size and reading the file.. But still you are right, will fix.